src/bridge/cockpitbridge.c

```diff
@@ -0,0 +1,271 @@
+
+  object_class = G_OBJECT_CLASS (class);
+  object_class->set_property = cockpit_bridge_set_property;
+  object_class->finalize = cockpit_bridge_finalize;
+
+  g_object_class_install_property (object_class, PROP_INIT_RECEIVED,
+                                   g_param_spec_boolean ("init-received", NULL,
+                                                         NULL, FALSE,
+                                                         G_PARAM_WRITABLE |
+                                                         G_PARAM_CONSTRUCT_ONLY |
+                                                         G_PARAM_STATIC_STRINGS));
+}
+
+static void
+cockpit_bridge_init (CockpitBridge *self)
+{
+  /* Owns the channels */
+  self->channels = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, g_object_unref);
+  self->payloads = g_hash_table_new_full (g_str_hash, g_str_equal, g_free, NULL);
+}
+
+static void
+on_channel_closed (CockpitChannel *channel,
+                   const gchar *problem,
+                   gpointer user_data)
+{
+  CockpitBridge *self = user_data;
+  g_hash_table_remove (self->channels, cockpit_channel_get_id (channel));
+}
+
+static void
+process_init (CockpitBridge *self,
+              CockpitTransport *transport,
+              JsonObject *options)
+{
+  gint64 version = -1;
+
+  if (!cockpit_json_get_int (options, "version", -1, &version))
+    {
+      g_warning ("invalid version field in init message");
+      cockpit_transport_close (transport, "protocol-error");
+    }
+
+  if (version == 1)
+    {
+      g_debug ("received init message");
+      self->init_received = TRUE;
+    }
+  else
+    {
+      g_message ("unsupported version of cockpit protocol: %" G_GINT64_FORMAT, version);
+      cockpit_transport_close (transport, "not-supported");
+    }
+}
+
+static void
+process_open (CockpitBridge *self,
+              CockpitTransport *transport,
+              const gchar *channel_id,
+              JsonObject *options)
+{
+  CockpitChannel *channel;
+  GType channel_type;
+  type_function channel_function = NULL;
+  const gchar *payload;
+
+  if (!channel_id)
+    {
+      g_warning ("Caller tried to open channel with invalid id");
+      cockpit_transport_close (transport, "protocol-error");
+    }
+  else if (g_hash_table_lookup (self->channels, channel_id))
+    {
+      g_warning ("Caller tried to reuse a channel that's already in use");
+      cockpit_transport_close (transport, "protocol-error");
+    }
+
+  else
+    {
+      if (!cockpit_json_get_string (options, "payload", NULL, &payload))
+        payload = NULL;
+
+      /* This will close with "not-supported" */
+      channel_type = COCKPIT_TYPE_CHANNEL;
+
+      channel_function = g_hash_table_lookup (self->payloads, payload);
+      if (channel_function != NULL)
+        {
+          channel_type = channel_function ();
+        }
+
+      g_debug ("Running channel type %lu for %s default is %lu", channel_type, payload, COCKPIT_TYPE_CHANNEL);
```

Comment:
If we want this message to be readable we should use ```g_type_name ()``` to format the types. Are you sure we need to list the default type?